src/network/networkd-dhcp6.c

```diff
@@ -76,6 +82,233 @@ static int dhcp6_lease_information_acquired(sd_dhcp6_client *client,
+}
+
+static Link *dhcp6_del_dynamic_delegation(struct in6_addr *addr) {
+        assert(dhcp6_prefixes);
+
+        return hashmap_remove(dhcp6_prefixes, addr);
+}
+
+static int dhcp6_del_dynamic_delegation_link(Link *link) {
+        Link *l;
+        struct in6_addr addr;
+        Iterator i;
+
+        assert(dhcp6_prefixes);
+        assert(link);
+
+        HASHMAP_FOREACH_KEY(l, addr, link->manager->links, i) {
+                if (link == l)
+                        dhcp6_del_dynamic_delegation(&addr);
+        }
+
+        if (hashmap_isempty(dhcp6_prefixes))
+                dhcp6_prefixes = hashmap_free(dhcp6_prefixes);
+
+        return 0;
+}
+
+static int dhcp6_pd_prefix_assign(Link *link, struct in6_addr *prefix,
+                                  uint8_t prefix_len,
+                                  uint32_t lifetime_preferred,
+                                  uint32_t lifetime_valid) {
+        sd_radv *radv = link->radv;
+        int r;
+        sd_radv_prefix *p;
+
+        r = sd_radv_prefix_new(&p);
+        if (r < 0)
+                goto error;
+
+        r = sd_radv_prefix_set_prefix(p, prefix, prefix_len);
+        if (r < 0)
+                goto error;
+
+        r = sd_radv_prefix_set_preferred_lifetime(p, lifetime_preferred);
+        if (r < 0)
+                goto error;
+
+        r = sd_radv_prefix_set_valid_lifetime(p, lifetime_valid);
+        if (r < 0)
+                goto error;
+
+        r = sd_radv_stop(radv);
+        if (r < 0)
+                goto error;
+
+        r = sd_radv_add_prefix(radv, p, true);
+        if (r < 0 && r != -EEXIST)
+                goto error;
+
+        r = sd_radv_start(radv);
+        if (r < 0)
+                goto error;
+
+        r = dhcp6_add_dynamic_delegation(&p->opt.in6_addr, link);
+        if (r < 0)
+                goto error;
+
+ error:
+        sd_radv_prefix_unref(p);
+
+        return r;
+}
+
+static Network *dhcp6_reset_pd_prefix_network(Link *link) {
+        assert(link);
+        assert(link->manager);
+        assert(link->manager->networks);
+
+        return link->manager->networks;
+}
+
+static int dhcp6_pd_prefix_distribute(Link *dhcp6_link, Iterator *i,
+                                      struct in6_addr *pd_prefix,
+                                      uint8_t pd_prefix_len,
+                                      uint32_t lifetime_preferred,
+                                      uint32_t lifetime_valid) {
+        Link *link;
+        Manager *manager = dhcp6_link->manager;
+        union in_addr_union prefix;
+        uint8_t n_prefixes, n_used = 0;
+        _cleanup_free_ char *buf = NULL;
+        int r;
+
+        assert(manager);
+        assert(pd_prefix_len <= 64);
+
+        prefix.in6 = *pd_prefix;
+
+        r  = in_addr_mask(AF_INET6, &prefix, pd_prefix_len);
+        if (r < 0)
+                return r;
+
+        n_prefixes = 1 << (pd_prefix_len - 64);
+
+        (void) in_addr_to_string(AF_INET6, &prefix, &buf);
+
```

Comment:
Better to remove this empty line. It suggests that those are two separate things, but it's really one.